packages/eslint-plugin/src/rules/explicit-function-return-type.ts

```diff
@@ -168,56 +187,68 @@ export default createRule<Options, MessageIds>({
     }
 
-    return {
-      'ArrowFunctionExpression, FunctionExpression'(
-        node: TSESTree.ArrowFunctionExpression | TSESTree.FunctionExpression,
-      ): void {
-        if (
-          options.allowConciseArrowFunctionExpressionsStartingWithVoid &&
-          node.type === AST_NODE_TYPES.ArrowFunctionExpression &&
-          node.expression &&
-          node.body.type === AST_NODE_TYPES.UnaryExpression &&
-          node.body.operator === 'void'
-        ) {
-          return;
-        }
+    function exitFunctionExpression(
+      node: TSESTree.ArrowFunctionExpression | TSESTree.FunctionExpression,
+    ): void {
+      const info = popFunctionInfo();
 
-        if (isAllowedFunction(node)) {
-          return;
-        }
+      if (
+        options.allowConciseArrowFunctionExpressionsStartingWithVoid &&
+        node.type === AST_NODE_TYPES.ArrowFunctionExpression &&
+        node.expression &&
+        node.body.type === AST_NODE_TYPES.UnaryExpression &&
+        node.body.operator === 'void'
+      ) {
+        return;
+      }
 
-        if (
-          options.allowTypedFunctionExpressions &&
-          (isValidFunctionExpressionReturnType(node, options) ||
-            ancestorHasReturnType(node))
-        ) {
-          return;
-        }
+      if (isAllowedFunction(node)) {
+        return;
+      }
 
-        checkFunctionReturnType(node, options, context.sourceCode, loc =>
-          context.report({
-            node,
-            loc,
-            messageId: 'missingReturnType',
-          }),
-        );
-      },
-      FunctionDeclaration(node): void {
+      if (
+        options.allowTypedFunctionExpressions &&
+        (isValidFunctionExpressionReturnType(node, options) ||
+          ancestorHasReturnType(node))
+      ) {
+        return;
+      }
+
+      checkFunctionReturnType(info, options, context.sourceCode, loc =>
+        context.report({
+          node,
+          loc,
+          messageId: 'missingReturnType',
+        }),
+      );
+    }
+
+    return {
+      'ArrowFunctionExpression, FunctionExpression, FunctionDeclaration':
+        enterFunction,
+      'ArrowFunctionExpression:exit': exitFunctionExpression,
+      'FunctionExpression:exit': exitFunctionExpression,
+      'FunctionDeclaration:exit'(node): void {
+        const info = popFunctionInfo();
         if (isAllowedFunction(node)) {
           return;
         }
         if (options.allowTypedFunctionExpressions && node.returnType) {
           return;
         }
 
-        checkFunctionReturnType(node, options, context.sourceCode, loc =>
+        checkFunctionReturnType(info, options, context.sourceCode, loc =>
           context.report({
             node,
             loc,
             messageId: 'missingReturnType',
           }),
         );
       },
+      ReturnStatement(node): void {
+        const current = functionInfoStack[functionInfoStack.length - 1];
+        current.returns.push(node);
+      },
```

Comment:
[Bug] Someone might write a `return` outside of a function. TypeScript will give an error, but the linter may still run. [[playground link](https://deploy-preview-8508--typescript-eslint.netlify.app/play/#ts=5.3.3&fileType=.ts&code=E4UwLgrsB2DcQ&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6RAD2MbMroDNYTMvkoB7JrWiJ8saJKKl0URNGhjokcGAC%2BIHUA&tsconfig=&tokens=false "standalone return in the typescript-eslint playground")]

Reply:
Thanks! I changed to use as assertion since without it,  `no-unnecessary-condition` reports lint errors